src/doc/src/reference/semver.md

```diff
@@ -1883,6 +1884,137 @@ Making a previously `unsafe` associated function or method on structs / enums
 safe is also a minor change, while the same is not true for associated
 function on traits (see [any change to trait item signatures](#trait-item-signature)).
 
+### Major: add a potentially shadowing method {#fn-add-potentially-shadowing-method}
+
+If you have a type which implements `Deref<Target=T>`, you must not add methods
+which may "shadow" methods in `T`. This can lead to unexpected changes in
+program behavior.
+
+```rust,ignore,run-fail
+// MAJOR CHANGE
+
+///////////////////////////////////////////////////////////
+// Before
+#[derive(Clone, Copy)]
+pub struct MySmartPtr<T>(pub T);
+
+impl<T> core::ops::Deref for MySmartPtr<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+///////////////////////////////////////////////////////////
+// After
+#[derive(Clone, Copy)]
+pub struct MySmartPtr<T>(pub T);
+
+impl<T> core::ops::Deref for MySmartPtr<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> MySmartPtr<T> {
+    pub fn method(self) -> usize {
+        2
+    }
+}
+
+///////////////////////////////////////////////////////////
+// Example usage that will break.
+use updated_crate::MySmartPtr;
+
+struct SomeStruct;
+
+impl SomeStruct {
+    fn method(&self) -> usize {
+        1
+    }
+}
+
+fn main() {
+    let mut ptr = MySmartPtr(SomeStruct);
+    assert_eq!(ptr.method(), 1);
+}
+```
+
+Note that the shadowing and shadowed methods receive `self`
+slightly differently: `&self` and `&mut self`.
+That's because Rust searches for methods first by value, then by `&`, then
+by `&mut T`, then by `*const T`. Rust stops the search
```

Comment:
Is this true that `*const T` is on the list?  That's not how we [document it](https://doc.rust-lang.org/nightly/reference/expressions/method-call-expr.html#method-call-expressions).

Reply:
Mention of `*const T` removed.

Reply:
Even if it isn't in the reference, it would be super useful to have a completely-accurate writeup somewhere, for the sake of cargo-semver-checks. We implement lints as best we understand the domain, and any divergences between what we implement and what rustc actually does are experienced as bugs by our users.

So personally, I'd advocate for either a special case callout in the reference ("if the `self` type is `*mut T` ..."), or at least a comment here containing the SemVer implications of such a case. Knowing about it now can save cargo-semver-checks maintainers hours of trying to triage bugs and figure out the exact rule later 🙏

Reply:
> In retrospect I was wrong to mention this here, and I'll remove it. It does slightly bother me that the reference is incomplete here, but as far as I know it always has been, and I assume somebody made a principled decision to keep it clear and comprehensible instead of describing every last detail.

That may or may not be true.  Typically we do want to describe every last user-visible behavioral detail unless we've decided to treat the behavior as not-guaranteed for some reason.  This doesn't strike me as the latter case offhand.  If the Reference isn't complete here, we probably just want to update it.

@ehuss might have some history and further context here.

Reply:
That is to say, @adetaylor, we'd welcome a PR to the Reference that would take a shot at describing this better.

Reply:
I had a crack at a major rewrite of that bit of the reference in https://github.com/rust-lang/reference/pull/1725.